lib/isaac_state_manager.go

```diff
@@ -0,0 +1,195 @@
+		},
+		common.BallotStateINIT,
+	)
+
+	for {
+		select {
+		case <-timer.C:
+			if sm.state.ballotState == common.BallotStateACCEPT {
+				sm.increaseRound()
+				break
+			}
+			go sm.broadcastExpiredBallot(sm.state)
+			sm.state.ballotState = sm.state.ballotState.Next()
+			sm.resetTimer(timer, sm.state.ballotState)
+
+		case state := <-sm.stateTransit:
+			switch state.ballotState {
+			case common.BallotStateINIT:
+				sm.proposeOrWait(timer, state)
+			case common.BallotStateSIGN:
+				sm.state = state
+				timer.Reset(sm.conf.TimeoutSIGN)
+			case common.BallotStateACCEPT:
+				sm.state = state
+				timer.Reset(sm.conf.TimeoutACCEPT)
+			case common.BallotStateALLCONFIRM:
+				sm.NextHeight()
+			case common.BallotStateNONE:
+				timer.Reset(sm.conf.TimeoutINIT)
+				log.Error("Wrong IsaacState", "IsaacState", state)
+			}
+
+		case <-sm.nextHeight:
+			round := sm.state.round
+			round.BlockHeight++
+			round.Number = 0
+			sm.TransitIsaacState(round, common.BallotStateINIT)
+
+		case <-sm.stop:
+			return
+		}
+	}
+}
+
+func (sm *IsaacStateManager) broadcastExpiredBallot(state IsaacState) {
+	round := round.Round{
+		Number:      state.round.Number,
+		BlockHeight: sm.nr.consensus.LatestConfirmedBlock.Height,
+		BlockHash:   sm.nr.consensus.LatestConfirmedBlock.Hash,
+		TotalTxs:    sm.nr.consensus.LatestConfirmedBlock.TotalTxs,
+	}
+
+	newExpiredBallot := NewBallot(sm.nr.localNode, round, []string{})
+	newExpiredBallot.SetVote(state.ballotState.Next(), common.VotingEXP)
+	newExpiredBallot.Sign(sm.nr.localNode.Keypair(), sm.nr.networkID)
+
+	sm.nr.ConnectionManager().Broadcast(*newExpiredBallot)
+}
+
+func (sm *IsaacStateManager) resetTimer(timer *time.Timer, state common.BallotState) {
+	switch state {
+	case common.BallotStateINIT:
+		timer.Reset(sm.conf.TimeoutINIT)
+	case common.BallotStateSIGN:
+		timer.Reset(sm.conf.TimeoutSIGN)
+	case common.BallotStateACCEPT:
+		timer.Reset(sm.conf.TimeoutACCEPT)
+	}
+}
+
+func (sm *IsaacStateManager) proposeOrWait(timer *time.Timer, state IsaacState) {
+	timer.Reset(time.Duration(1 * time.Hour))
+	proposer := sm.nr.CalculateProposer(state.round.BlockHeight, state.round.Number)
+	log.Debug("calculated proposer", "proposer", proposer)
+
+	if proposer == sm.nr.localNode.Address() {
+		if err := sm.nr.proposeNewBallot(state.round.Number); err == nil {
+			log.Debug("propose new ballot", "proposer", proposer, "round", state.round, "ballotState", common.BallotStateSIGN)
+			sm.state = state
+			sm.state.ballotState = common.BallotStateSIGN
+			timer.Reset(sm.conf.TimeoutSIGN)
+		} else {
+			log.Error("failed to proposeNewBallot", "height", sm.nr.consensus.LatestConfirmedBlock.Height, "error", err)
+			sm.state = state
+			timer.Reset(sm.conf.TimeoutINIT)
+		}
+	} else {
+		sm.state = state
+		timer.Reset(sm.conf.TimeoutINIT)
+	}
+}
+
+func (sm *IsaacStateManager) State() IsaacState {
+	return sm.state
+}
+
+func (sm *IsaacStateManager) Stop() {
+	go func() {
+		sm.stop <- true
+	}()
+}
```

Comment:
What is use-cases that use `Stop`?   If the most case is that node is stopping,  the node/server should wait to stop `IsaacStateManage`?  

IMHO,  It can handle `IsaacStateManger` lifecycle in `main()` with `rungroup` (not node_runner) (It's just another option :-)

Reply:
@anarcher `Stop` can be used only for testing at this time, but it will he useful :)